lib/chef/resource/windows_share.rb

```diff
@@ -331,8 +330,35 @@ def bool_string(bool)
           # bool ? 1 : 0
           bool ? "$true" : "$false"
         end
+
+        # Grant-SmbShareAccess sets only one permission to one user so we need to remove from the lower permission access
+        # For change_users remove common full_users from it
+        # For read_users remove common full_users as well as change_users
+        def new_resource_users
+          @full_users = new_resource.full_users
+          @change_users = new_resource.change_users - new_resource.full_users
+          @read_users = new_resource.read_users - (new_resource.full_users + new_resource.change_users)
+        end
```

Comment:
Change the instance variables underneath the property is not only a bad idea, I'm fairly certain it's the reasons for the issues you're seeing.

Instead, do something like:

```ruby
def effective_full_users
  new_resource.full_users
end

def effective_change_users
  new_resource.change_users - new_resource.full_users
end

def effective_read_users
  new_resource.read_users - (new_resource.full_users + new_resource.change_users)
end
```

Then instead of using `instance_variable_get`, you can call `new_resource.send`.

This is the moral equivalent of what @tas50 suggested back in 2020 (though with a method instead of a var, he said "use a different variable"), you said it didn't work, but you didn't say what specifically didn't work, and it should definitely work. Lets do that and then debug it, because otherwise we're setting up a footgun here.